tests/smoke/alp_zgesvd.cpp

```diff
@@ -0,0 +1,276 @@
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#ifdef _COMPLEX
+#include <complex>
+#include <cmath>
+#include <iomanip>
+#endif
+
+#include <alp.hpp>
+#include <alp/algorithms/svd.hpp>
+#include <graphblas/utils/iscomplex.hpp> // use from grb
+#ifdef DEBUG
+#include "../utils/print_alp_containers.hpp"
+#endif
+
+using namespace alp;
+
+using BaseScalarType = double;
+using Orthogonal = structures::Orthogonal;
+using General = structures::General;
+
+#ifdef _COMPLEX
+using ScalarType = std::complex< BaseScalarType >;
+#else
+using ScalarType = BaseScalarType;
+#endif
+
+constexpr BaseScalarType tol = 1.e-10;
+constexpr size_t RNDSEED = 1;
+
+//** generate random rectangular matrix data: complex version */
+template< typename T >
+std::vector< T > generate_rectangular_matrix_data(
+	size_t N,
+	size_t M,
+	const typename std::enable_if<
+		grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * M );
+	for( size_t i = 0; i < N; ++i ) {
+		for( size_t j = 0; j < M; ++j ) {
+			T val( std::rand(), std::rand() );
+			data[ i * M + j ] = val / std::abs( val );
+		}
+	}
+	return data;
+}
+
+//** generate random rectangular matrix data: real version */
+template<
+	typename T
+>
+std::vector< T >  generate_rectangular_matrix_data(
+	size_t N,
+	size_t M,
+	const typename std::enable_if<
+		!grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * M );
+	for( size_t i = 0; i < N; ++i ) {
+		for( size_t j = 0; j < M; ++j ) {
+			data[ i * M + j ] = static_cast< T >( std::rand() ) / RAND_MAX;
+		}
+	}
+	return data;
+}
+
+template<
+	typename D,
```

Comment:
As done in other algorithms, since most of the containers' template arguments are not relevant I would access the publicly accessible ones via member types (eg, MatrixT::structure, etc..)